File: 해시/leetcode496(Easy)-Next Greater Element I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        index_table = collections.defaultdict(int)
        for i, num in enumerate(nums2):
            index_table[num] = i

        def search(target_number, start_index):
            for i in range(start_index, len(nums2)):
                if nums2[i] > target_number:
                    return nums2[i]
            return -1

        ans = []
        for num in nums1:
            index = index_table[num]
            value  = search(num, index + 1)
            ans.append(value)
            
        return ans

File: 해시/leetcode496(Easy)-Next Greater Element I.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  const indexTable = nums2.reduce((obj, num, index) => {
    obj[num] = index
    return obj
  }, {})

  const search = (targetNumber, startIndex) => {
    for (let i = startIndex; i < nums2.length; i += 1) {
      if (nums2[i] > targetNumber) {
        return nums2[i]
      }
    }
    return -1
  }

  return nums1.map((num) => {
    const index = indexTable[num]
    return search(num, index + 1)
  })
}
